### PR TITLE
feat: remapping contexts

### LINKED ISSUE #1855: Support context aware remappings from nested foundry imports
### Component

Forge

### Describe the feature you would like

Currently forge picks up remappings from imported libraries as well, if a nested library and the project use the same remapping but for different targets (different version of the same dependency for example), the current behaviour is that the project's remapping overrides the nested lib's remapping, which may not be intended.

solc remappings support a `context:` prefix to circumvent this:

https://docs.soliditylang.org/en/latest/path-resolution.html#import-remapping

`context:prefix=target`

where `context` must match the beginning of the source unit name of the file containing the import.

so that solc can be invoked like

```console
solc module1:github.com/ethereum/dapp-bin/=dapp-bin/ \
     module2:github.com/ethereum/dapp-bin/=dapp-bin_old/ \
     source.sol
```

This would require some upstream changes in ethers-rs:

- [ ] add optional context field to `Remapping`

this would also require us to track where the Remapping is coming from so we can derive the `context` (which should be the relative path `lib/<dep>/` I think)

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by onbjerg**
## Motivation

Dependencies in projects using Foundry can mess with how imports are resolved through the project because remappings from dependencies are just merged with remappings defined in the root folder.

To solve this, we make use of solc's remapping contexts.

## Solution

This PR adds support for specifying remapping contexts if necessary, but ideally it should never be. Instead, this PR also automatically scopes each remapping defined in dependencies to files in the dependency itself.

Global remappings (`context: None`) are still automatically generated for each installed dependency, making it possible to import them as usual.

This PR in other words just makes the following user-facing change: remappings defined in dependencies/libraries can no longer override remappings defined in `foundry.toml`/`remappings.txt`, and it should no longer be necessary to copy over remappings from dependencies to `foundry.toml`/`remappings.txt` for dependencies with weird remappings.

This PR also adds a `--pretty` flag to `forge remappings` that groups remappings by context. Global remappings apply to *every file* (including dependencies), and remappings with a context only apply to the paths defined by the context. The default is to *not* pretty print since a common use case of `forge remappings` is to pipe the output to a file or other command.

## Todo

- [ ] Add tests
- [x] Merge ethers PR
- [x] Remove patch

Closes https://github.com/foundry-rs/foundry/issues/3440, closes #1855

Depends on https://github.com/gakonst/ethers-rs/pull/2509

### CODE REVIEW

**Comment by gakonst** [comment]
Ethers 2.0.8 is out with the fix.

**Comment by aathan** [comment]
Some comments/questions: 
(1) How does this actually operate? are the remappings loaded from dependencies internally contextualized (with context:) for a given forge run, or are you re-writing the remappings files written to the lib/ projects when those are deposited? I assume the former.
(2) is there an opinion as to whether there should be an *option* to also scope the top level/global remappings so that they can *optionally* only apply to the "main" project files, and never to any libs that may have some import that happens to match? E.g., is it possible for a dependent project to itself depend on the same other project you also depend on, and in that case, could the global remappings end up affecting the dependency's imports?
(3) does someone know definitively, to tell me, whether the way solc resolves type names may cause structs of the same name, declared at the file-scope in various dependencies and/or the root project, to collide? If the solc devs need a poke about that, I can go open a feature request.

File: cli/src/cmd/forge/remappings.rs [hunk 1]
```diff
@@ -14,15 +16,46 @@ pub struct RemappingArgs {
...
+    /// Pretty-print the remappings, grouping each of them by context.
+    #[clap(long)]
+    pretty: bool,
 }
 impl_figment_convert_basic!(RemappingArgs);
 
 impl Cmd for RemappingArgs {
     type Output = ();
 
+    // TODO: Do people use `forge remappings >> file`?
```
Comment (Evalir) [inline cli/src/cmd/forge/remappings.rs:28]: How are we planning to test this? Wasn't this what you tweeted the other day?

File: config/src/providers/remappings.rs [hunk 2]
```diff
@@ -83,11 +90,13 @@ impl<'a> RemappingsProvider<'a> {
         new_remappings.extend(remappings);
 
         // scan all library dirs and autodetect remappings
+        // todo: if a lib specifies contexts for remappings manually, we need to figure out how to
```
Comment (Evalir) [inline config/src/providers/remappings.rs:93]: do we plan to do this now or later?

File: config/src/providers/remappings.rs [hunk 3]
```diff
@@ -152,8 +165,16 @@ impl<'a> RemappingsProvider<'a> {
...
                 }
 
-                let mut remappings =
-                    config.remappings.into_iter().map(|m| m.into()).collect::<Vec<Remapping>>();
+                let mut remappings = config
+                    .remappings
+                    .into_iter()
+                    .map(Remapping::from)
+                    .map(|mut r| {
+                        // todo: windows/mac/linux
```
Comment (Evalir) [inline config/src/providers/remappings.rs:173]: 😄

**Review by Evalir — COMMENTED** [review]
So this seems to work well, and considering this will fix master i'm leaning to merging soon—some comments:

File: Cargo.toml [hunk 4]
```diff
@@ -61,16 +61,16 @@ panic = "abort"
...
+ethers = { version = "2.0.8", default-features = false }
+ethers-addressbook = { version = "2.0.8", default-features = false }
+ethers-core = { version = "2.0.8", default-features = false }
+ethers-contract = { version = "2.0.8", default-features = false }
+ethers-contract-abigen = { version = "2.0.8", default-features = false }
+ethers-providers = { version = "2.0.8", default-features = false }
+ethers-signers = { version = "2.0.8", default-features = false }
+ethers-middleware = { version = "2.0.8", default-features = false }
+ethers-etherscan = { version = "2.0.8", default-features = false }
+ethers-solc = { version = "2.0.8", default-features = false }
```
Comment (mattsse) [inline Cargo.toml:73]: now this is on ethers master, we should continue using git deps

File: cli/src/cmd/forge/remappings.rs (hunk 1 above)
Comment (mattsse) [inline cli/src/cmd/forge/remappings.rs:28]: they definitely do, 
I also used this a few times at least

File: config/src/providers/remappings.rs [hunk 5]
```diff
@@ -47,8 +47,15 @@ impl<'a> RemappingsProvider<'a> {
         trace!("get all remappings from {:?}", self.root);
         /// prioritizes remappings that are closer: shorter `path`
         ///   - ("a", "1/2") over ("a", "1/2/3")
-        fn insert_closest(mappings: &mut HashMap<String, PathBuf>, key: String, path: PathBuf) {
-            match mappings.entry(key) {
+        /// grouped by remapping context
+        fn insert_closest(
+            mappings: &mut HashMap<Option<String>, HashMap<String, PathBuf>>,
```
Comment (mattsse) [inline config/src/providers/remappings.rs:52]: I think this should work,
so the previous hashmap is now just the `None` key

and this also supports remappings with context extracted from nested deps

**Review by mattsse — APPROVED** [review]
cool, this looks less invasive than I expected, and since context is optional, this should work as before

smol nit re deps, otherwise lgtm

File: config/src/providers/remappings.rs (hunk 5 above)
Comment (Evalir) [reply config/src/providers/remappings.rs:52]: fixed in https://github.com/foundry-rs/foundry/pull/5397/commits/d92bd235fa551d073b7bfb09a3011a7d264387f5 — we can switch to master deps whenever we plan on publishing crates 😄

File: Cargo.toml (hunk 4 above)
Comment (Evalir) [reply Cargo.toml]: fixed in https://github.com/foundry-rs/foundry/pull/5397/commits/d92bd235fa551d073b7bfb09a3011a7d264387f5 — we can switch to master deps whenever we plan on publishing crates 😄

**Pull request merged**